File: Part8/Fortran/FunctionConcepts/CompilerErrors.f90
! Fortran Function Errors - All errors active for compilation

program FortranFunctionErrors
    implicit none
    
    ! Error: Missing 'end subroutine' statement
    subroutine printHello()
        print *, "Hello"
    
    ! Error: Wrong end identifier (doesn't match subroutine name)
    subroutine sayHi()
        print *, "Hi"
    end subroutine greet
    
    ! Error: Using 'end' without specifying 'subroutine' or 'function'
    subroutine display()
        print *, "Display"
    end
    
    ! Error: Missing return type in function declaration
    function getValue()
        getValue = 42
    end function getValue
    
    ! Error: Missing 'result' clause but assigning to wrong variable
    function getNumber() result(num)
        getNumber = 5
    end function getNumber
    
    ! Error: Using 'return' with value (Fortran doesn't work this way)
    integer function calculate()
        return 42
    end function calculate
    
    ! Error: Missing function assignment
    real function getPrice()
        real :: price
        price = 15.99
    end function getPrice
    
    ! Error: Using 'void' or 'none' instead of subroutine
    void printMessage()
        print *, "Message"
    end void printMessage
    
    none doSomething()
        print *, "Something"
    end none doSomething
    
    ! Error: Using 'call' with a function (should only call subroutines)
    integer function add(a, b)
        integer, intent(in) :: a, b
        add = a + b
    end function add
    
    subroutine testCall()
        integer :: result
        call add(5, 10)
    end subroutine testCall
    
    ! Error: Missing 'call' keyword for subroutine
    subroutine process()
        print *, "Processing"
    end subroutine process
    
    subroutine execute()
        process()
    end subroutine execute
    
    ! Error: Using parentheses in subroutine call with no arguments
    subroutine show()
        print *, "Showing"
    end subroutine show
    
    subroutine run()
        call show()
    end subroutine run
    
    ! Error: Missing comma between parameters
    subroutine setPizza(size toppings)
        character(len=*), intent(in) :: size
        integer, intent(in) :: toppings
        print *, size, toppings
    end subroutine setPizza
    
    ! Error: Using semicolon instead of comma between parameters
    real function pizzaCost(size; toppings)
        character(len=*), intent(in) :: size
        integer, intent(in) :: toppings
        pizzaCost = 15.99
    end function pizzaCost
    
    ! Error: Missing intent specification (not an error but bad practice, but wrong syntax here)
    subroutine updateValue(count in out)
        integer :: count
        count = 5
    end subroutine updateValue
    
    ! Error: Using 'default' for optional parameters
    subroutine orderPizza(size, toppings default 2)
        character(len=*), intent(in) :: size
        integer, intent(in) :: toppings
        print *, "Order placed"
    end subroutine orderPizza
    
    ! Error: Named arguments using := instead of =
    subroutine makePizza(size, count)
        character(len=*), intent(in) :: size
        integer, intent(in) :: count
        print *, size, count
    end subroutine makePizza
    
    subroutine placeOrder()
        call makePizza(size := "large", count := 3)
    end subroutine placeOrder
    
    ! Error: Using 'begin' instead of no keyword after declaration
    subroutine calculate() begin
        integer :: x
        x = 5
    end subroutine calculate
    
    ! Error: Missing parameter type declarations inside subroutine
    subroutine getValue(number)
        print *, number
    end subroutine getValue
    
    ! Error: Function name doesn't match in 'end function'
    integer function factorial(n)
        integer, intent(in) :: n
        if (n <= 1) then
            factorial = 1
        else
            factorial = n * factorial(n - 1)
        end if
    end function calculate
    
    ! Error: Using 'procedure' keyword incorrectly
    procedure myProc()
        print *, "Procedure"
    end procedure myProc
    
    ! Error: Parameter type after name instead of with double colon
    subroutine setData(count integer, price real)
        count = 5
        price = 10.0
    end subroutine setData
    
    ! Error: Using 'with parameters' syntax
    subroutine doTask()
        print *, "Task"
    end subroutine doTask
    
    subroutine runTask()
        call doTask with parameters()
    end subroutine runTask

end program FortranFunctionErrors

File: Part8/C#/FunctionConcepts/CompilerErrors.cs
// C# Function Errors - All errors active for compilation

using System;

class CSharpFunctionErrors
{
    // Error: Missing return type
    PrintHello()
    {
        Console.WriteLine("Hello");
    }

    // Error: Missing semicolon after statement in method
    void SayHi()
    {
        Console.WriteLine("Hi")
    }

    // Error: Using 'function' keyword instead of method declaration
    function void Greet()
    {
        Console.WriteLine("Greetings");
    }

    // Error: Missing return statement in non-void method
    int GetNumber()
    {
        int x = 5;
    }

    // Error: Returning value from void method
    void ProcessData()
    {
        return 42;
    }

    // Error: Return type doesn't match returned value
    int GetValue()
    {
        return 3.14;
    }

    // Error: Using 'call' keyword before method name
    void MyMethod()
    {
        Console.WriteLine("Test");
    }

    void Test()
    {
        call MyMethod();
    }

    // Error: Missing parentheses in method call
    void Display()
    {
        Console.WriteLine("Display");
    }

    void Run()
    {
        Display;
    }

    // Error: Using 'none' instead of 'void'
    none PrintMessage()
    {
        Console.WriteLine("Message");
    }

    // Error: Parameter type after parameter name
    void SetValue(count int, price float)
    {
        Console.WriteLine($"{count} {price}");
    }

    // Error: Missing parameter type
    float CalculateTotal(size, int toppings)
    {
        return 15.99f;
    }

    // Error: Using 'default' keyword incorrectly for default parameter
    void OrderPizza(string size default "medium", int toppings)
    {
        Console.WriteLine("Order placed");
    }

    // Error: Default parameter before required parameter
    void MakePizza(string size = "large", int toppings, bool delivery = false)
    {
        Console.WriteLine("Making pizza");
    }

    // Error: Using := for named arguments
    void SetPizza(string size, int count)
    {
        Console.WriteLine($"{size} {count}");
    }

    void Order()
    {
        SetPizza(size := "large", count := 3);
    }

    // Error: Using semicolon instead of comma between parameters
    void Configure(int width; int height; bool fullscreen)
    {
        Console.WriteLine("Configured");
    }

    // Error: Missing comma between parameters in definition
    int AddNumbers(int a int b)
    {
        return a + b;
    }

    // Error: Using 'begin' and 'end' instead of braces
    void Calculate() begin
        int x = 5;
    end

    // Error: Semicolon after method header before opening brace
    void ProcessValue(); 
    {
        int x = 10;
    }

    // Error: Wrong number of arguments in method call
    int Add(int a, int b)
    {
        return a + b;
    }

    void CallAdd()
    {
        int result = Add(5);
    }

    // Error: Using 'with parameters' syntax
    void DoSomething()
    {
        Console.WriteLine("Done");
    }

    void Execute()
    {
        call DoSomething with parameters ();
    }

    // Error: Parameter name in method call without colon
    void CreateOrder(string product, int quantity)
    {
        Console.WriteLine($"{product}: {quantity}");
    }

    void PlaceOrder()
    {
        CreateOrder(product = "Pizza", quantity = 2);
    }

    // Error: Missing parentheses in method declaration
    void ShowMessage
    {
        Console.WriteLine("Message");
    }

    // Error: Using 'procedure' instead of 'void'
    procedure DisplayInfo()
    {
        Console.WriteLine("Info");
    }

    // Error: Using 'return' outside of method
    int globalValue = 10;
    return globalValue;

    static void Main(string[] args)
    {
        Console.WriteLine("Main");
    }
}
